extensions/amp-ad-network-doubleclick-impl/0.1/sra-utils.js

```diff
@@ -295,3 +302,69 @@ function serializeItem_(key, value) {
     (Array.isArray(value) ? value : [value]).map(encodeURIComponent).join();
   return `${encodeURIComponent(key)}=${serializedValue}`;
 }
+
+/**
+ * Callback for streaming SRA response given creative and JSON parsed Object
+ * containing headers (as if request had been sent via non-SRA flow).  Creative
+ * and headers object is converted to XHR FetchResponse object and passed to
+ * resolver popped off stack of resolvers (in order of expected block responses)
+ * such that sendXhrRequest is resolved from standard A4A flow.  Done boolean
+ * used to verify array of resolvers is empty once all results are returned.
+ * @param {string} creative
+ * @param {!Object<string,string>} headersObj
+ * @param {boolean} done
+ * @param {!Array<function(?../../../src/service/xhr-impl.FetchResponse)>} sraRequestAdUrlResolvers
+ * @param {string} sraUrl url of SRA request for error reporting
+ */
+export function sraBlockCallbackHandler(
+  creative, headersObj, done, sraRequestAdUrlResolvers, sraUrl) {
+  const headerNames = Object.keys(headersObj);
+  if (headerNames.length == 1 &&
+      isObject(headersObj[headerNames[0]])) {
+    // TODO(keithwrightbos) - fix upstream so response does
+    // not improperly place headers under key.
+    headersObj =
+      /** @type {!Object} */(headersObj)[headerNames[0]];
+    headersObj = Object.keys(headersObj).reduce(
+        (newObj, key) => {
+          newObj[key.toLowerCase()] = headersObj[key];
+          return newObj;
+        }, {});
+  }
+  // Force safeframe rendering method.
+  headersObj[RENDERING_TYPE_HEADER.toLowerCase()] =
+      XORIGIN_MODE.SAFEFRAME;
+  // Construct pseudo fetch response to be passed down the A4A
+  // promise chain for this block.
+  const headers =
+/** @type {?../../../src/service/xhr-impl.FetchResponseHeaders} */
```

Comment:
Nit: indent is off here and below in various places.

Reply:
This is how the linter wants it... go figure.